protocol/src/packet.rs

```diff
@@ -0,0 +1,190 @@
+use std::io::{self, Read, Write};
+
+use flate2::read::ZlibDecoder;
+use flate2::write::ZlibEncoder;
+use flate2::Compression;
+use minecraft_protocol_derive::{Decoder, Encoder};
+
+use crate::decoder::{Decoder, DecoderReadExt};
+use crate::encoder::{Encoder, EncoderWriteExt};
+use crate::error::{DecodeError, EncodeError};
+
+#[derive(Debug, Clone)]
+pub struct Packet {
+    pub id: i32,
+    pub data: Vec<u8>,
+}
+
+impl Packet {
+    pub fn encode<W: Write>(
+        &self,
+        writer: &mut W,
+        compression_threshold: Option<i32>,
+    ) -> Result<(), EncodeError> {
+        let mut buf = Vec::new();
+        let packet = RawPacket {
+            id: self.id,
+            data: self.data.clone(),
+        };
+        if let Some(threshold) = compression_threshold {
+            CompressedRawPacket { packet, threshold }.encode(&mut buf)?;
+        } else {
+            packet.encode(&mut buf)?;
+        }
+
+        writer.write_var_i32(buf.len() as i32)?;
+        writer.write_all(&buf)?;
+
+        Ok(())
+    }
+
+    pub fn decode<R: Read>(reader: &mut R, compressed: bool) -> Result<Packet, DecodeError> {
+        let len = match reader.read_var_i32() {
+            Ok(len) => len as usize,
+            Err(DecodeError::IoError { io_error })
+                if io_error.kind() == io::ErrorKind::UnexpectedEof =>
+            {
+                return Err(DecodeError::Incomplete { bytes_needed: 1 })
+            }
+            Err(err) => return Err(err.into()),
+        };
+
+        let mut buf = Vec::with_capacity(len);
+        let bytes_read = reader.take(len as u64).read_to_end(&mut buf)?;
+
+        if bytes_read != len {
+            return Err(DecodeError::Incomplete {
+                bytes_needed: len - bytes_read,
+            });
+        }
+
+        let RawPacket { id, data } = if compressed {
+            CompressedRawPacket::decode(&mut buf.as_slice())?
+        } else {
+            RawPacket::decode(&mut buf.as_slice())?
+        };
+
+        Ok(Self { id, data })
+    }
+}
+
+#[derive(Debug, Clone, Encoder, Decoder)]
+struct RawPacket {
+    #[data_type(with = "var_int")]
+    pub id: i32,
+    #[data_type(with = "rest")]
+    pub data: Vec<u8>,
+}
+
+#[derive(Debug, Clone)]
+struct CompressedRawPacket {
+    packet: RawPacket,
+    threshold: i32,
+}
+
+impl Encoder for CompressedRawPacket {
+    fn encode<W: Write>(&self, writer: &mut W) -> Result<(), EncodeError> {
+        let mut packet_buf = Vec::new();
+        self.packet.encode(&mut packet_buf)?;
+
+        let data_len = packet_buf.len() as i32;
+        if self.threshold >= 0 && data_len > self.threshold {
+            writer.write_var_i32(data_len)?;
+            let mut encoder = ZlibEncoder::new(writer, Compression::default());
+            encoder.write_all(&packet_buf)?;
+            encoder.finish()?;
+        } else {
+            writer.write_var_i32(0)?;
```

Comment:
Why zero?

Reply:
https://wiki.vg/Protocol#With_compression

> If the size of the buffer containing the packet data and ID (as a VarInt) is smaller than the threshold specified in the packet Set Compression. It will be sent as uncompressed. This is done by setting the data length as 0. (Comparable to sending a non-compressed format with an extra 0 between the length, and packet data).